test/Transactions/Orleans.Transactions.Tests/Grains/MultiStateTransactionalGrains.cs

```diff
@@ -0,0 +1,144 @@
+    {
+        public SingleStateTransactionalGrain(
+            [TransactionalState("data", TransactionTestConstants.TransactionStore)] ITransactionalState<GrainData> data,
+            ILoggerFactory loggerFactory)
+            :base(new ITransactionalState<GrainData>[1]{data}, loggerFactory)
+        {
+        }
+    }
+
+    public class MultiStateTransactionalGrainBaseClass: Grain, ITransactionTestGrain
+    {
+        protected ITransactionalState<GrainData>[] dataArray;
+        private readonly ILoggerFactory loggerFactory;
+        protected ILogger logger;
+
+        public MultiStateTransactionalGrainBaseClass(
+            ITransactionalState<GrainData>[] dataArray,
+            ILoggerFactory loggerFactory)
+        {
+            this.dataArray = dataArray;
+            this.loggerFactory = loggerFactory;
+        }
+
+        public override Task OnActivateAsync()
+        {
+            this.logger = this.loggerFactory.CreateLogger(this.GetGrainIdentity().ToString());
+            return base.OnActivateAsync();
+        }
+
+        public Task Set(int newValue)
+        {
+            foreach (var data in this.dataArray)
+            {
+                TransactionalGrainUtils.Set(data, newValue, this.logger);
+            }
+            return Task.CompletedTask;
+        }
+
+        public Task<int[]> Add(int numberToAdd)
+        {
+            return Task.FromResult(this.dataArray.Select(data => TransactionalGrainUtils.Add(data, numberToAdd, this.logger)).ToArray());
+        }
+
+        public Task<int[]> Get()
+        {
+            return Task.FromResult(this.dataArray.Select(data => TransactionalGrainUtils.Get(data, this.logger)).ToArray());
+        }
+
+        public Task AddAndThrow(int numberToAdd)
+        {
+            foreach (var data in dataArray)
+            {
+                TransactionalGrainUtils.Add(data, numberToAdd, this.logger);
+            }
+            throw new Exception($"{GetType().Name} test exception");
+        }
+
+        public Task Deactivate()
+        {
+            DeactivateOnIdle();
+            return Task.CompletedTask;
+        }
+    }
+
+    public static class TransactionalGrainUtils
+    {
+        public static void Set(ITransactionalState<GrainData> data, int newValue, ILogger logger)
+        {
+            logger.LogInformation($"Setting from {data.State.Value} to {newValue}.");
+            data.State.Value = newValue;
+            data.Save();
+            logger.LogInformation($"Set to {data.State.Value}.");
+        }
+
+        public static int Get(ITransactionalState<GrainData> data, ILogger logger)
+        {
+            logger.LogInformation($"Get {data.State.Value}.");
+            return data.State.Value;
+        }
+
+        public static int Add(ITransactionalState<GrainData> data, int numberToAdd, ILogger logger)
+        {
+            logger.LogInformation($"Adding {numberToAdd} to {data.State.Value}.");
+            data.State.Value += numberToAdd;
```

Comment:
In general, we should use a non-commutative operation (eg subtraction) for transaction tests in order to help ensure that the order of application is correct.

It doesn't matter so much here since this isn't a correctness test (or are they all correctness tests?)

Reply:
why is this ? did I fail my highschool math test again?

Reply:
It's fine the way it is. It's just that we want to verify some of the properties of transactions (ACID, etc) and one way to verify some of those properties is to verify that the order of application is consistent. One way to verify order is to use a non-commutative operation on the value here.

LGTM, just want to give Jason a chance to look at it.

Reply:
I know it looks fine now since we don't have correctness tests. Was just asking for more info on non-commutative operation and how it verify the order of application. I can totally search it, but I guess it is faster to ask you,.

Reply:
and @jason-bragg can you take a look of this pr ?

Reply:
We'll likely need to adjust the operation logic when we add correctness testing, but accounting for this is, imo, out of scope for this PR.

Side note: non-commutative is insufficient, as it, to my recollection, concerns itself with the order of the datum within an operation, not the order of operations.  We need something that detects missing or out of order operations.   That is 100 - 1 - 2 == 100 - 2 - 1,  which doesn't help us.

Reply:
yes that was my confusion, since subtraction doesn't 100% verify the order either, so I was wondering what's other magic there I don't know of